firedrake/preconditioners/pmg.py

```diff
@@ -274,55 +314,259 @@ def __init__(self, Vf, Vc, Vf_bcs, Vc_bcs):
+
+        m = mx;  k = nx;  n = ny*nz*nel;
+        lda = (tflag>0)? nx : mx;
+        dgemm_(&TA1, &notr, &m,&n,&k, &one, A1,&lda, x,&k, &zero, y,&m);
+
+        p = 0;  s = 0;
+        m = mx;  k = ny;  n = my;
+        lda = (tflag>0)? ny : my;
+        for(int i=0; i<nz*nel; i++){
+           dgemm_(&notr, &TA2, &m,&n,&k, &one, y+p,&m, A2,&lda, &zero, x+s,&m);
+           p += m*k;
+           s += m*n;
+        }
+
+        p = 0;  s = 0;
+        m = mx*my;  k = nz;  n = mz;
+        lda = (tflag>0)? nz : mz;
+        for(int i=0; i<nel; i++){
+           dgemm_(&notr, &TA3, &m,&n,&k, &one, x+p,&m, A3,&lda, &zero, y+s,&m);
+           p += m*k;
+           s += m*n;
+        }
+        return;
+        }
+        """
+
+        # UFL elements order the component DoFs related to the same node contiguously.
+        # We transpose before and after the multiplcation times J to have each component
+        # stored contiguously as a scalar field, thus reducing the number of dgemm calls.
 
+        # We could benefit from loop tiling for the transpose, but that makes the code
+        # more complicated.
+
+        prolong_code = f"""
+        {kronmxv_code}
+
+        void prolongation(double *restrict y, const double *restrict x){{
+            double JX[{mx}*{nx}] = {{ {JX} }};
+            double t0[{lwork}], t1[{lwork}];
+            double one=1.0E0;
+
+            for(int j=0; j<{nxyz}; j++)
+                for(int i=0; i<{nscal}; i++)
+                    t0[j + {nxyz}*i] = x[i + {nscal}*j];
+
+            kronmxv(0, {mx},{my},{mz}, {nx},{ny},{nz}, {nscal}, JX,{JY},{JZ}, t0,t1);
+
+            for(int j=0; j<{mxyz}; j++)
+                for(int i=0; i<{nscal}; i++)
+                   y[i + {nscal}*j] = t1[j + {mxyz}*i];
+            return;
+        }}
+        """
+
+        restrict_code = f"""
+        {kronmxv_code}
+
+        void restriction(double *restrict y, const double *restrict x,
+        const double *restrict w){{
+            double JX[{mx}*{nx}] = {{ {JX} }};
+            double t0[{lwork}], t1[{lwork}];
+            double one=1.0E0;
+
+            for(int j=0; j<{mxyz}; j++)
+                for(int i=0; i<{nscal}; i++)
+                    t0[j + {mxyz}*i] = x[i + {nscal}*j] * w[i + {nscal}*j];
+
+            kronmxv(1, {nx},{ny},{nz}, {mx},{my},{mz}, {nscal}, JX,{JY},{JZ}, t0,t1);
+
+            for(int j=0; j<{nxyz}; j++)
+                for(int i=0; i<{nscal}; i++)
+                    y[i + {nscal}*j] += t1[j + {nxyz}*i];
+            return;
+        }}
+        """
+        self.prolong_kernel = op2.Kernel(prolong_code, "prolongation", ldargs=["-lblas"])
+        self.restrict_kernel = op2.Kernel(restrict_code, "restriction", ldargs=["-lblas"])
+
+    @staticmethod
+    def get_nodes_1d(V):
+        # Return GLL nodes if V==CG or GL nodes if V==DG
+        from FIAT import quadrature
+        from FIAT.reference_element import DefaultLine
+        use_tensorproduct, N, family = tensor_product_space_query(V)
```

Comment:
I guess one should also assert that the `variant` of the ufl_element is `"spectral"` not `"equispaced"`?

Reply:
since spectral is the default, I am just checking if someone is using equispaced points by mistake